src/rust/src/lib.rs

```diff
@@ -66,8 +73,130 @@ mod rust_bridge {
 mod b64;
 use b64::{from_base64, to_base64};
 
-mod contract;
-use contract::invoke_host_function;
-
 mod log;
 use crate::log::init_logging;
+
+// The tree of dependencies for the contract host is multi-versioned to reduce
+// the possibility of observable semantic drift over code versions.
+//
+// Formally: assume ledger N is originally executed on-chain by copes of core
+// all running against code version X. Then later assume that we wish to upgrade
+// dependencies to code version Y. We say that drift has occurred if N replays
+// with a different tx result or meta hash under X and Y. We want to ensure that
+// if there is drift between X and Y when replaying N, we catch and compensate
+// for it, either by preserving code version X in core forever and replaying N
+// under it, or at least keeping around a special-case path in code version Y
+// for the narrow difference between X and Y that is exercised by the
+// transactions in N, along with code to use this special-case path when
+// replaying N.
+//
+// The strategy we use to avoid this is similar to that of protocol upgrades, in
+// the sense that we will provide an identifier for the version of code
+// transactions are executing under, and use consensus 'upgrades' to schedule
+// synchronous transition across the network from one such version to the next.
+//
+// The difference with protocol versions is that clients of the network should
+// in general be oblivious to such changes -- the goal is to _avoid_ observable
+// differences between changes -- and so we will not necessarily announce them
+// publicly or require explicit votes from operators: anyone can attempt such an
+// upgrade at any time and any node capable of supporting such an upgrade will
+// vote for it automatically; if sufficient support exists the network will
```

Comment:
What does "automatically" means here? Will a validator be able to audit the new code before approving the upgrade?

Reply:
Validators never deploy new versions of core automatically; this is about whether _upgrading core_ is enough of a signal that you're willing to use new code, or whether you want to hold back and have a _subsequent vote_ to activate it. I'll clarify the language here (and make it clearer this is a maybe-policy question rather than a definite plan).